File: frontend/src/api/draft.ts
export interface DraftSettings {
  id: number
  platform: string
  platform_draft_id: string | null
  league_id: number | null
  season: string
  format: string
  num_teams: number
  num_rounds: number
  my_slot: number
  rank_set_id: number | null
  roster_positions: string[] | null
  team_names: Record<string, string>
}

export interface PickRow {
  pick_number: number
  round: number
  slot: number
  platform_player_id: string
  name: string
  position: string | null
  team: string | null
}

export interface DraftStatus {
  draft: DraftSettings
  picks: PickRow[]
  next_pick_number: number | null
  current_round: number | null
  current_slot: number | null
  is_my_turn: boolean
  is_complete: boolean
}

export interface QueueRow {
  platform_player_id: string
  name: string
  position: string | null
  team: string | null
}

export interface CreateDraftParams {
  season: string
  format: string
  num_teams: number
  num_rounds: number
  my_slot: number
}

export interface CreateSleeperDraftParams {
  platform_draft_id: string
  format: string
  my_slot: number
}

export interface CreateDraftFromLeagueParams {
  league_id: number
  my_slot: number
}

/** A lightweight per-draft summary from `GET /drafts` -- not a full
 * DraftStatus (no picks, no rank_set_id/roster_positions). Used by the
 * Leagues list/detail views to show "is there a draft, and how far along."
 */
export interface DraftListRow {
  id: number
  platform: string
  league_id: number | null
  season: string
  format: string
  num_teams: number
  num_rounds: number
  my_slot: number
  pick_count: number
  next_pick_number: number | null
  current_round: number | null
  is_complete: boolean
  created_at: string
}

const API_BASE = import.meta.env.VITE_API_BASE ?? 'http://127.0.0.1:8000'

/** Thrown by parseOrThrow with the real HTTP status attached, so a caller can
 * distinguish "not found" (safe to fall back to a fresh setup flow) from a
 * transient failure (should surface as an error, not silently invite creating
 * a duplicate draft).
 */
export class ApiError extends Error {
  status: number

  constructor(message: string, status: number) {
    super(message)
    this.status = status
  }
}

async function parseOrThrow<T>(response: Response, label: string): Promise<T> {
  if (!response.ok) {
    let detail = ''
    try {
      const body = (await response.json()) as { detail?: string }
      detail = body.detail ?? ''
    } catch {
      // response body wasn't JSON -- fall back to just the status code
    }
    throw new ApiError(
      detail ? `${label}: ${detail}` : `${label} failed: ${response.status}`,
      response.status,
    )
  }
  return response.json() as Promise<T>
}

export async function createDraft(
  params: CreateDraftParams,
): Promise<DraftStatus> {
  const response = await fetch(`${API_BASE}/drafts`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  return parseOrThrow(response, 'Creating draft')
}

export async function createSleeperDraft(
  params: CreateSleeperDraftParams,
): Promise<DraftStatus> {
  const response = await fetch(`${API_BASE}/drafts/sleeper`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  return parseOrThrow(response, 'Creating Sleeper draft')
}

export async function createDraftFromLeague(
  params: CreateDraftFromLeagueParams,
): Promise<DraftStatus> {
  const response = await fetch(`${API_BASE}/drafts/league`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  return parseOrThrow(response, 'Creating draft from league')
}

export async function fetchDrafts(leagueId?: number): Promise<DraftListRow[]> {
  const query = leagueId != null ? `?league_id=${leagueId}` : ''
  const response = await fetch(`${API_BASE}/drafts${query}`)
  return parseOrThrow(response, 'Fetching drafts')
}

export async function fetchDraftStatus(draftId: number): Promise<DraftStatus> {
  const response = await fetch(`${API_BASE}/drafts/${draftId}`)
  return parseOrThrow(response, 'Fetching draft')
}

export async function syncSleeperDraft(draftId: number): Promise<DraftStatus> {
  const response = await fetch(`${API_BASE}/drafts/${draftId}/sync`, {
    method: 'POST',
  })
  return parseOrThrow(response, 'Syncing draft')
}

export async function switchToManual(draftId: number): Promise<DraftStatus> {
  const response = await fetch(
    `${API_BASE}/drafts/${draftId}/switch-to-manual`,
    {
      method: 'POST',
    },
  )
  return parseOrThrow(response, 'Switching to manual')
}

export async function makePick(
  draftId: number,
  platformPlayerId: string,
): Promise<{ pick_number: number }> {
  const response = await fetch(`${API_BASE}/drafts/${draftId}/picks`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ platform_player_id: platformPlayerId }),
  })
  return parseOrThrow(response, 'Making pick')
}

export async function undoLastPick(
  draftId: number,
): Promise<{ pick_number: number } | null> {
  const response = await fetch(`${API_BASE}/drafts/${draftId}/picks`, {
    method: 'DELETE',
  })
  return parseOrThrow(response, 'Undoing pick')
}

export async function fetchQueue(draftId: number): Promise<QueueRow[]> {
  const response = await fetch(`${API_BASE}/drafts/${draftId}/queue`)
  return parseOrThrow(response, 'Fetching queue')
}

export async function saveQueue(
  draftId: number,
  platformPlayerIds: string[],
): Promise<{ count: number }> {
  const response = await fetch(`${API_BASE}/drafts/${draftId}/queue`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ platform_player_ids: platformPlayerIds }),
  })
  return parseOrThrow(response, 'Saving queue')
}

File: backend/app/api/draft.py
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.db import get_db
from app.draft import (
    DraftError,
    create_draft,
    create_draft_from_league,
    create_sleeper_draft,
    get_status,
    list_drafts,
    list_queue,
    make_pick,
    replace_queue,
    switch_to_manual,
    sync_sleeper_draft,
    undo_last_pick,
)

router = APIRouter(prefix="/drafts")

DbSession = Annotated[Session, Depends(get_db)]


class CreateDraftRequest(BaseModel):
    season: str
    format: str
    num_teams: int
    num_rounds: int
    my_slot: int


class CreateSleeperDraftRequest(BaseModel):
    platform_draft_id: str
    format: str
    my_slot: int


class CreateDraftFromLeagueRequest(BaseModel):
    league_id: int
    my_slot: int


class DraftSummary(BaseModel):
    id: int
    platform: str
    platform_draft_id: str | None
    league_id: int | None
    season: str
    format: str
    num_teams: int
    num_rounds: int
    my_slot: int
    rank_set_id: int | None
    roster_positions: list[str] | None
    team_names: dict[str, str]


class DraftListRow(BaseModel):
    """A lightweight per-draft summary for the Leagues UI -- not a full
    DraftStatus (no picks, no rank_set_id/roster_positions -- those need a
    League join that list_drafts() deliberately skips per-row to avoid N+1)."""

    id: int
    platform: str
    league_id: int | None
    season: str
    format: str
    num_teams: int
    num_rounds: int
    my_slot: int
    pick_count: int
    next_pick_number: int | None
    current_round: int | None
    is_complete: bool
    created_at: datetime


class PickRow(BaseModel):
    pick_number: int
    round: int
    slot: int
    platform_player_id: str
    name: str
    position: str | None
    team: str | None


class DraftStatus(BaseModel):
    draft: DraftSummary
    picks: list[PickRow]
    next_pick_number: int | None
    current_round: int | None
    current_slot: int | None
    is_my_turn: bool
    is_complete: bool


class MakePickRequest(BaseModel):
    platform_player_id: str


class MakePickResponse(BaseModel):
    pick_number: int


class QueueRow(BaseModel):
    platform_player_id: str
    name: str
    position: str | None
    team: str | None


class ReplaceQueueRequest(BaseModel):
    platform_player_ids: list[str]


class ReplaceQueueResponse(BaseModel):
    count: int


@router.get("", response_model=list[DraftListRow])
def get_drafts(db: DbSession, league_id: int | None = None) -> list[DraftListRow]:
    return [DraftListRow(**row) for row in list_drafts(db, league_id)]


@router.post("", response_model=DraftStatus)
def post_draft(payload: CreateDraftRequest, db: DbSession) -> DraftStatus:
    draft = create_draft(
        db, payload.season, payload.format, payload.num_teams, payload.num_rounds, payload.my_slot
    )
    status = get_status(db, draft.id)
    return DraftStatus(**status)


@router.post("/sleeper", response_model=DraftStatus)
def post_sleeper_draft(payload: CreateSleeperDraftRequest, db: DbSession) -> DraftStatus:
    try:
        draft = create_sleeper_draft(db, payload.platform_draft_id, payload.format, payload.my_slot)
    except DraftError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    status = get_status(db, draft.id)
    return DraftStatus(**status)


@router.post("/league", response_model=DraftStatus)
def post_draft_from_league(payload: CreateDraftFromLeagueRequest, db: DbSession) -> DraftStatus:
    try:
        draft = create_draft_from_league(db, payload.league_id, payload.my_slot)
    except DraftError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    status = get_status(db, draft.id)
    return DraftStatus(**status)


@router.post("/{draft_id}/sync", response_model=DraftStatus)
def post_sync(draft_id: int, db: DbSession) -> DraftStatus:
    try:
        status = sync_sleeper_draft(db, draft_id)
    except DraftError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return DraftStatus(**status)


@router.post("/{draft_id}/switch-to-manual", response_model=DraftStatus)
def post_switch_to_manual(draft_id: int, db: DbSession) -> DraftStatus:
    try:
        status = switch_to_manual(db, draft_id)
    except DraftError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return DraftStatus(**status)


@router.get("/{draft_id}", response_model=DraftStatus)
def get_draft_status(draft_id: int, db: DbSession) -> DraftStatus:
    status = get_status(db, draft_id)
    if status is None:
        raise HTTPException(status_code=404, detail="Draft not found")
    return DraftStatus(**status)


@router.post("/{draft_id}/picks", response_model=MakePickResponse)
def post_pick(draft_id: int, payload: MakePickRequest, db: DbSession) -> MakePickResponse:
    try:
        result = make_pick(db, draft_id, payload.platform_player_id)
    except DraftError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return MakePickResponse(**result)


@router.delete("/{draft_id}/picks", response_model=MakePickResponse | None)
def delete_last_pick(draft_id: int, db: DbSession) -> MakePickResponse | None:
    try:
        undone = undo_last_pick(db, draft_id)
    except DraftError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if undone is None:
        return None
    return MakePickResponse(pick_number=undone["pick_number"])


@router.get("/{draft_id}/queue", response_model=list[QueueRow])
def get_queue(draft_id: int, db: DbSession) -> list[QueueRow]:
    return [QueueRow(**row) for row in list_queue(db, draft_id)]


@router.put("/{draft_id}/queue", response_model=ReplaceQueueResponse)
def put_queue(draft_id: int, payload: ReplaceQueueRequest, db: DbSession) -> ReplaceQueueResponse:
    count = replace_queue(db, draft_id, payload.platform_player_ids)
    return ReplaceQueueResponse(count=count)
